base/reduce.jl

```diff
@@ -296,46 +296,95 @@ function extrema(itr)
     return (vmin, vmax)
 end
 
-## all & any
+## any
 
-function mapfoldl(f, ::AndFun, itr)
+any(itr)    = any(IdFun(), itr)
+any(f, itr) = any(specialized_unary(f), itr)
+
+function any(f::IdFun, itr)
     for x in itr
-        !f(x) && return false
+        x && return true
     end
-    return true
+    return false
 end
 
-function mapfoldl(f, ::OrFun, itr)
+function any(f::Func{1}, itr)
     for x in itr
         f(x) && return true
     end
     return false
 end
 
-function mapreduce_impl(f, op::AndFun, A::AbstractArray{Bool}, ifirst::Int, ilast::Int)
-    while ifirst <= ilast
-        @inbounds x = A[ifirst]
-        !f(x) && return false
-        ifirst += 1
+function any(f::IdFun, itr::Array{Bool, 1})
+    i   = 1
+    len = length(itr)
+    while i <= len
+       @inbounds x = itr[i]
```

Comment:
I'd think `@inbounds for x in itr` should be equally fast. `@inbounds` is necessary for optimal performance due to #11350, but whether we should actually use it for non-Arrays is another question, since we'll segfault on user-defined types if `done` is incorrectly defined.